CC13XX CC26XX : transition from pinmux to pinctrl

vaishnavachath opened the pull request:
This pull request contains the changes to migrate from pinmux to pinctrl API for CC13XX and CC26XX based platforms as outlined in https://github.com/zephyrproject-rtos/zephyr/issues/39740 .
The pull request contains the following changes in the respective commits:

1.  Introduces the pinctrl driver for CC13XX/CC26XX platform.
2. migrate existing consumers to use the new driver, pinctrl API.
3. remove obsolete pinmux driver.
4. fix white-space/styling issues for defines moved from hal/driverlib/ioc.h - intentionally kept as separate commit for now, can be squashed with the first commit  if that is preferred.

The change-set was tested on CC1352R sensor-tag to verify the I2C/SPI/UART IOC functionality is not broken.

vaishnavachath commented:
@cfriedt @bwitherspoon Can you please review the PR?

cfriedt reviewed (dismissed):
LGTM - thanks so much for getting these changes in!

@vaishnavachath - would be great if you could convince TI to allow you to be platform maintainer for Zephyr 😃

gmarull commented on drivers/pinctrl/Kconfig.cc13xx_cc26xx:
```diff
@@ -0,0 +1,11 @@
+	depends on SOC_SERIES_CC13X2_CC26X2
+	default $(dt_compat_enabled,$(DT_COMPAT_CC13XX_CC26XX_PINCTRL))
+	help
+	  Enable the TI SimpleLink CC13xx / CC26xx pinctrl driver
```
nit: missing newline at the eof

gmarull commented on dts/bindings/pinctrl/ti,cc13xx-cc26xx-pinctrl.yaml:
```diff
@@ -0,0 +1,28 @@
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+    TI SimpleLink CC13xx / CC26xx pinctrl node.
+
+    Device pin configuration should be placed in the child nodes of this node.
+    Populate the 'pinmux' field with a pair consisting of a pin number and its IO
+    function
```
please extend this documentation with some examples, supported properties, etc. See other vendors as an example.

vaishnavachath replied:
added documentation

gmarull commented on include/dt-bindings/pinctrl/cc13xx_cc26xx-pinctrl.h:
```diff
@@ -0,0 +1,159 @@
+//
+// IOC Peripheral Port Mapping
+//
+//*****************************************************************************
```
please no C99 style comments

vaishnavachath replied:
this file is a trimmed down version from hal/ti/..driverlib/ioc.h , all styling issues fixed for this file as separate commit

gmarull commented on include/dt-bindings/pinctrl/cc13xx_cc26xx-pinctrl.h:
```diff
@@ -0,0 +1,159 @@
+                                 IOC_IOPULL_UP | IOC_SLEW_DISABLE |                 \
+                                 IOC_HYST_DISABLE | IOC_NO_EDGE |                   \
+                                 IOC_INT_DISABLE | IOC_IOMODE_OPEN_DRAIN_NORMAL |   \
+                                 IOC_NO_WAKE_UP | IOC_INPUT_ENABLE )                 
```
standard properties need to be used. For example, `bias-pull-up`. This is encoding pin properties (bias, pull-up, slew-rate, etc.) in a custom way. Ref. https://docs.zephyrproject.org/latest/guides/pinctrl/index.html#pin-control-drivers

vaishnavachath replied:
implemented standard properties

gmarull commented on boards/arm/cc1352r1_launchxl/cc1352r1_launchxl-pinctrl.dtsi:
```diff
@@ -0,0 +1,45 @@
+#include <dt-bindings/pinctrl/cc13xx_cc26xx-pinctrl.h>
+
+&pinctrl {
+    /* UART0 */
+    uart0_tx_default: uart0_tx_default {
+		pinmux = <13 IOC_PORT_MCU_UART0_TX IOC_STD_OUTPUT>;
```
nit: please don't mix tabs/spaces

mbolivar-nordic reviewed (commented):
Other than @gmarull's comment on missing examples and details, the bindings changes LGTM.

cfriedt commented:
@vaishnavachath - can you make the changes requested by @gmarull ?

vaishnavachath committed:
drivers: pinctrl: add CC13XX/CC26XX pinctrl driver

Add pinctrl driver for CC13XX/CC26XX family of SoCs
to facilitate transition from pinmux to pinctrl.

`IOCPortConfigureSet()` from TI hal driverlib used to
implement the generic pinctrl driver.

Signed-off-by: Vaishnav Achath <vaishnav@beagleboard.org>

vaishnavachath committed:
drivers: CC1XX/CC26XX based boards: transition to pinctrl driver

This commit has the necessary changes to update the consumers
of pinmux driver(SPI, I2C, UART) and update the board specific
files to use the pinctrl interface.

Signed-off-by: Vaishnav Achath <vaishnav@beagleboard.org>

vaishnavachath committed:
drivers: pinmux: remove cc13xx_cc26xx pinmux driver

all the consumers of the obsolete pinmux driver is
updated to use pinctrl API, this commit removes
the pinmux driver and assosciated sections.

Signed-off-by: Vaishnav Achath <vaishnav@beagleboard.org>

vaishnavachath commented:
@gmarull @cfriedt @mbolivar-nordic , I have made the suggested changes, can you please review the updates.
Changes made:
 - implement standard properties (bias-pull-up, bias-pull-down .etc)
 - update DT binding documentation with examples
 - earlier the styling fixes on driverlib header was separate commit which created confusion, that was fixed in the initial commit itself

mbolivar-nordic reviewed (dismissed):
Bindings LGTM. The rest I only reviewed quickly, but it all seems reasonable as well. Thanks for the update!

cfriedt commented:
@vaishnavachath - looks like some minor changes to get the last two tests to pass. Are you running with Twister?

gmarull commented on drivers/pinctrl/pinctrl_cc13xx_cc26xx.c:
```diff
@@ -0,0 +1,34 @@
+int pinctrl_configure_pins(const pinctrl_soc_pin_t *pins, uint8_t pin_cnt, uintptr_t reg)
+{
+	ARG_UNUSED(reg);
+	int i;
+
+	for (i = 0; i < pin_cnt; i++) {
```
```suggestion
	for (uint8_t i = 0U; i < pin_cnt; i++) {
```

vaishnavachath replied:
Updated with the suggested change

gmarull reviewed (dismissed):
LGTM, one minor non-blocking nitpick

vaishnavachath commented:
> @vaishnavachath - looks like some minor changes to get the last two tests to pass. Are you running with Twister?

@cfriedt made small mistake in cc26xx dtsi while updating which is fixed now and tests are passing.
@gmarull updated with suggested minor change in `pinctrl_cc13xx_cc26xx.c`
